### PR TITLE
extend ShapedOreRecipe along with implementing a fluid ingredient for crafting recipes

### PR DESCRIPTION
**Pull request by PrototypeTrousers**
## What
Allows crafting recipes to actually CHECK for the actual fluid in a bucket. since Forge doesnt account for any of it by itself.

## Implementation Details
Properly implementing container item methods allows for use of drums and quantum tanks to be used along with those recipes, without being consumed

## Outcome
Fixes: #1125

## Potential Compatibility Issues
None *wishful thinking*

### CODE REVIEW

File: src/main/java/gregtech/api/recipes/ModHandler.java [hunk 1]
```diff
@@ -303,7 +304,7 @@ public static void addShapedRecipe(boolean withUnificationData, String regName,
             return;
         }
 
-        IRecipe shapedOreRecipe = new ShapedOreRecipe(null, result.copy(), finalizeShapedRecipeInput(recipe))
+        IRecipe shapedOreRecipe = new GTShapedOreRecipe(null, result.copy(), finalizeShapedRecipeInput(recipe))
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/api/recipes/ModHandler.java:307]: This should also be applied in `MirroredShaped`

File: src/main/java/gregtech/common/crafting/GTFluidCraftingIngredient.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    @Override
+    public boolean apply(@Nullable ItemStack testedStack) {
+        IFluidHandlerItem handler = null;
+        if (testedStack.hasCapability(CapabilityFluidHandler.FLUID_HANDLER_ITEM_CAPABILITY, null)) {
+            handler = testedStack.getCapability(CapabilityFluidHandler.FLUID_HANDLER_ITEM_CAPABILITY, null);
+        }
+        if (handler != null) {
+            if (fluidStack == null || fluidStack.amount == 0) return true;
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/common/crafting/GTFluidCraftingIngredient.java:26]: Should this return false?

File: src/main/java/gregtech/api/items/metaitem/stats/ItemFluidContainer.java [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+import net.minecraftforge.fluids.capability.IFluidHandlerItem;
+
+public class ItemFluidContainer implements IItemContainerItemProvider {
+    @Override
+    public ItemStack getContainerItem(ItemStack itemStack) {
+        IFluidHandlerItem handler = itemStack.getCapability(CapabilityFluidHandler.FLUID_HANDLER_ITEM_CAPABILITY, null);
+        if (handler != null) {
+            FluidStack drained = handler.drain(1000, false);
+            if (drained == null) return ItemStack.EMPTY;
+            if (drained.amount == 1000) {
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/api/items/metaitem/stats/ItemFluidContainer.java:15]: You have added this to cells that can contain greater than 1000 mb of fluid, as well as contain variable amounts of fluid, maybe being less than 1000 mb. Should this instead check if `drained.amount > 0`? and should drained be changed to `handler.drain(Integer.MAX_VALUE, false)`?

File: src/main/java/gregtech/api/block/machines/MachineItemBlock.java [hunk 4]
```diff
@@ -79,6 +82,31 @@ public ICapabilityProvider initCapabilities(@Nonnull ItemStack stack, @Nullable
...
+    }
+
+    @Override
+    public @Nonnull ItemStack getContainerItem(@Nonnull ItemStack itemStack) {
+        if (!hasContainerItem(itemStack)) {
+            return ItemStack.EMPTY;
+        }
+        if (itemStack.hasCapability(CapabilityFluidHandler.FLUID_HANDLER_ITEM_CAPABILITY, null)) {
+            IFluidHandlerItem handler = itemStack.getCapability(CapabilityFluidHandler.FLUID_HANDLER_ITEM_CAPABILITY, null);
+            if (itemStack.getCapability(CapabilityFluidHandler.FLUID_HANDLER_ITEM_CAPABILITY, null) != null) {
```
Comment (TechLord22) [inline src/main/java/gregtech/api/block/machines/MachineItemBlock.java:97]: This if statement should be removed.

File: src/main/java/gregtech/api/block/machines/MachineItemBlock.java [hunk 5]
```diff
@@ -79,6 +82,31 @@ public ICapabilityProvider initCapabilities(@Nonnull ItemStack stack, @Nullable
...
+                if (handler != null) {
+                    FluidStack drained = handler.drain(1000, true);
+                    if (drained == null || drained.amount != 1000) {
+                        return ItemStack.EMPTY;
+                    }
+                    return handler.getContainer().copy();
+                }
+            }
+        }
+        return itemStack.getCapability(CapabilityFluidHandler.FLUID_HANDLER_ITEM_CAPABILITY, null).getContainer().copy();
```
Comment (TechLord22) [inline src/main/java/gregtech/api/block/machines/MachineItemBlock.java:107]: Does this have a risk of a NullPointerException?

File: src/main/java/gregtech/common/crafting/GTFluidCraftingIngredient.java [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+import net.minecraftforge.fluids.capability.IFluidHandlerItem;
+
+import javax.annotation.Nullable;
+
+public class GTFluidCraftingIngredient extends Ingredient {
+    private final FluidStack fluidStack;
+
+    GTFluidCraftingIngredient(ItemStack... stacks) {
+        super(stacks);
+        this.fluidStack = stacks[0].getCapability(CapabilityFluidHandler.FLUID_HANDLER_ITEM_CAPABILITY, null).drain(Integer.MAX_VALUE, false);
```
Comment (TechLord22) [inline src/main/java/gregtech/common/crafting/GTFluidCraftingIngredient.java:16]: Should we throw an IllegalArgumentException here when the first ItemStack does not have the capability?

File: src/main/java/gregtech/common/crafting/GTShapelessOreRecipe.java [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+import net.minecraftforge.oredict.OreIngredient;
+import net.minecraftforge.oredict.ShapelessOreRecipe;
+
+import javax.annotation.Nonnull;
+
+public class GTShapelessOreRecipe extends ShapelessOreRecipe {
+    public GTShapelessOreRecipe(ResourceLocation group, @Nonnull ItemStack result, Object... recipe) {
+        super(group, result);
+        for (Object in : recipe)
+        {
```
Comment (TechLord22) [inline src/main/java/gregtech/common/crafting/GTShapelessOreRecipe.java:22]: This class needs the EditorConfig formatting applied.

**Review by TechLord22 — CHANGES_REQUESTED** [review]

**Review by TechLord22 — APPROVED** [review]

**Pull request merged**